pintc/src/predicate/optimize/const_folding.rs

```diff
@@ -0,0 +1,52 @@
+use fxhash::FxHashMap;
+
+use crate::{
+    error::Handler,
+    expr::{evaluate::Evaluator, Expr},
+    predicate::{Contract, ExprKey},
+    span::Spanned,
+    types::Type,
+};
+
+/// In a given contract, simplify all sub-expressions that evaluate to constants.
+pub(crate) fn const_folding(contract: &mut Contract) {
+    fold_consts(contract);
+}
+
+pub(crate) fn fold_consts(contract: &mut Contract) {
+    let mut replace_map: FxHashMap<ExprKey, (Expr, Type)> = FxHashMap::default();
+
+    let evaluator = Evaluator::new(&contract.enums);
+
+    for pred_key in contract.preds.keys().collect::<Vec<_>>() {
+        for expr_key in contract.exprs(pred_key) {
+            if let Ok(imm) = evaluator.evaluate_key(&expr_key, &Handler::default(), contract) {
+                // check if the imm is different or not -- how to do?
+                // currently replacing every single expr even if it is the same -- not ideal, lots of unnecessary dupes
+
+                let simplified_expr = Expr::Immediate {
+                    value: imm.clone(),
+                    span: expr_key.get(contract).span().clone(),
+                };
+                replace_map.insert(
+                    expr_key,
+                    (simplified_expr, expr_key.get_ty(contract).clone()),
+                );
+            }
```

Comment:
@mohammadfawaz Is there a way to tell if the new `imm` is a fold of the original `expr`? I've been struggling to think of a way to accomplish this check.

If we don't have a check here, then we replace every `expr` even if it doesn't need it. So the `replace_map` essentially duplicates everything in `contract.exprs`. Then it goes through the work of replacing every single `expr` even if we don't need to make a single replacement.

ex. 

```
predicate Test {
    var a = 1 + 2;
}

replacing --- 
old expr: 2
with
new expr: 2
--- 
replacing --- 
old expr: {0}
with
new expr: {0}
--- 
replacing --- 
old expr: 1
with
new expr: 1
--- 
replacing --- 
old expr: 0
with
new expr: 0
--- 
replacing --- 
old expr: (1 + 2)
with
new expr: 3
--- 
```

only the last `expr` above needs to be replaced.

A really dirty way to accomplish this would be to string check 
```
contract.with_ctrct(old_expr_key), != contract.with_ctrct(simplified_expr_key),
```

which I don't believe we'd be alright doing.

Reply:
Added. Seems to work great! Thanks @otrho